File: Min Stack.java
//with 2 stacks, TC-(1) sc-O(1)
// class MinStack {
//     Stack<Integer> st;
//     Stack<Integer> minSt;
//     int min;

//     public MinStack(){
//         this.st = new Stack<>();
//         this.minSt = new Stack<>();
//         this.min = Integer.MAX_VALUE;
//         minSt.push(min);
//     }
    
//     public void push(int val) {
//         min = Math.min(min, val);
//         st.push(val);
//         minSt.push(min);
  
//         }

    
//     public void pop() {
//         st.pop();
//         minSt.pop();
//         min = minSt.peek();
    
//     }
    
//     public int top() {
//         return st.peek();
       
//     }
    
//     public int getMin() {
//        return min;
//     }
// }

// with one stack

class MinStack {
    Stack<Integer> st;
    int min;

    public MinStack(){
        this.st = new Stack<>();
        this.min = Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        if(min >= val){
            st.push(min);
            min = val;
        }
        st.push(val);
  
        }

    
    public void pop() {
        int el = st.pop();
        if(el == min){
            min = st.pop();
        }
    
    }
    
    public int top() {
        return st.peek();
       
    }
    
    public int getMin() {
       return min;
    }
}
 /* class ListNode {
 *     int val, min;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
           min = x;
 *         next = null;
 *     }
 */ 

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: MinStack.cs
#TC:O(1)
public class MinStack {

    private Stack<int> st;
    private Stack<int> minSt;
    private int min;

    public MinStack() {
        this.min = int.MaxValue;
        this.st = new Stack<int>();
        this.minSt = new Stack<int>();
        this.minSt.Push(this.min);
    }

    public void Push(int val) {
        min = Math.Min(min, val);
        st.Push(val);
        minSt.Push(min);
    }

    public void Pop() {
        st.Pop();
        minSt.Pop();
        min = minSt.Peek();
    }

    public int Top() {
        return st.Peek();
    }

    public int GetMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */